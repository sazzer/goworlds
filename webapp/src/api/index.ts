import uriParser from 'uri-template';
import axios, {AxiosRequestConfig, AxiosResponse, Method} from 'axios';

/** The Base URI for the API */
const API_URI = process.env.REACT_APP_API_URI;
/**
 * The type of a Request to make
 */
export type Request = {
    method?: Method,
    urlParams?: any,
    body?: any,
    clientId?: string,
};

/**
 * The type of a Response to return from a request
 */
export type Response<T> = {
    status: number,
    body: T,
};

/**
 * Actually make a request to the service
 * @param url the URL to request
 * @param params the request parameters to provide
 */
export async function request<T>(url: string, params: Request = {}) : Promise<Response<T>> {
    /** The expanded URL */
    const expandedUrl = uriParser.parse(url).expand(params.urlParams);

    /** The parameters to pass to Axios */
    const fetchParams: AxiosRequestConfig = {
        method : params.method || 'GET',
        baseURL: API_URI,
        headers: {},
        data: params.body,
    };

    if (params.clientId) {
        fetchParams.auth = {
            username: params.clientId,
            password: ''
        };
    }

    try {
        const httpResponse: AxiosResponse = await axios(expandedUrl, fetchParams);

        return {
            status: httpResponse.status,
            body: httpResponse.data,
        }
    } catch (e) {
        console.log('Error making HTTP request: %s', url, e);
        throw e;
    }
}
