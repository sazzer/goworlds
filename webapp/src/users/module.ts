/** The module prefix */
export const MODULE_PREFIX = 'Users/';
