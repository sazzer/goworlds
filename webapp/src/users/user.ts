/** The shape of a user */
export declare type User = {
    id: string,
    created: string,
    updated: string,
    name: string,
    email: string,
};