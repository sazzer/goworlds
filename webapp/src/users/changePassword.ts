//////// Action for changing the users password

import {MODULE_PREFIX} from "./module";
import {
    Action,
    asyncAction,
    buildAction,
    buildSaga,
    failedAction,
    ResolvedAsyncAction,
    succeededAction
} from "../redux";
import {request} from "../api";
import {User} from "./user";

/** The action type for changing the users password */
export const CHANGE_PASSWORD_ACTION = MODULE_PREFIX + 'changePassword';

/** The shape of the action to change the users password */
declare type ChangePasswordAction = {
    userId: string,
    password: string,
    callback: (err: Error | undefined) => void
}

/**
 * Build the Redux action to change the users password
 * @param userId The ID of the User to update
 * @param password The new password
 * @param callback A callback to trigger when the action is completed
 */
export function changePassword(userId: string, password: string,
                                  callback?: (err: Error | undefined) => void): Action<ChangePasswordAction> {
    return buildAction(CHANGE_PASSWORD_ACTION, {
        userId,
        password,
        callback: callback || (() => {})
    });
}

/**
 * Saga to change a users password
 * @param action the action to handle
 */
export function* changePasswordSaga(action: Action<ChangePasswordAction>) : IterableIterator<any> {
    yield asyncAction(CHANGE_PASSWORD_ACTION, async () => {
        const response = await request<any>('/users/{id}', {
            method: 'PATCH',
            urlParams: {
                id: action.payload.userId
            },
            body: {
                password: action.payload.password,
            }
        });

        return {
            ...response.body,
        };
    }, action.payload);
}

/**
 * Saga for when changing a users password was a success
 * @param action the action to handle
 */
export function changePasswordSuccessSaga(action: ResolvedAsyncAction<ChangePasswordAction, User>) {
    if (action.input) {
        action.input.callback(undefined);
    }
}

/**
 * Saga for when changing a users password was a failure
 * @param action the action to handle
 */
export function changePasswordFailedSaga(action: ResolvedAsyncAction<ChangePasswordAction, Error>) {
    if (action.input) {
        action.input.callback(action.payload);
    }
}

//////// The actual module definitions

/** The sagas for this module */
export const sagas = [
    buildSaga(CHANGE_PASSWORD_ACTION, changePasswordSaga),
    buildSaga(succeededAction(CHANGE_PASSWORD_ACTION), changePasswordSuccessSaga),
    buildSaga(failedAction(CHANGE_PASSWORD_ACTION), changePasswordFailedSaga),
];
