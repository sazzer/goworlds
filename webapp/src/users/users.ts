import {
    Action, AsyncAction,
    asyncAction,
    buildAction,
    buildSaga,
    buildSelector, finishedAction,
    ResolvedAsyncAction,
    succeededAction
} from "../redux";
import {request} from "../api";
import {MODULE_PREFIX} from "./module";
import {createReducer} from "redux-create-reducer";
import produce from "immer";
import {select} from 'redux-saga/effects';
import {UPDATE_PROFILE_ACTION} from "./updateProfile";
import {User} from "./user";

/** The shape of the state for this module */
export declare type State = {
    users: Array<User>,
};

/** The initial state for this module */
const INITIAL_STATE: State = {
    users: [],
};

//////// Action for loading a user

/** The action type for registering a new user */
const LOAD_USER_ACTION = MODULE_PREFIX + 'loadUser';

/** The shape of the action to load a User by ID */
declare type LoadUserAction = {
    userId: string,
    force: boolean,
    callback: () => void
}

/**
 * Build the Redux action to load a user
 * @param userId The ID of the User to load
 * @param force Set to true to re-load the user even if we've already got a copy of it
 * @param callback Callback to trigger on loading the user
 */
export function loadUser(userId: string, force: boolean = false, callback?: () => void): Action<LoadUserAction> {
    return buildAction(LOAD_USER_ACTION, {
        userId,
        force,
        callback: callback || (() => {}),
    });
}

/**
 * Saga to load a user
 * @param action the action to handle
 */
export function* loadUserSaga(action: Action<LoadUserAction>) : IterableIterator<any> {
    let doAction = action.payload.force;
    if (!doAction) {
        const existingUser = yield select(selectUserById(action.payload.userId));
        doAction = existingUser === undefined;
    }

    if (doAction) {
        yield asyncAction(LOAD_USER_ACTION, async () => {
            const response = await request<any>('/users/{id}', {
                method: 'GET',
                urlParams: {
                    id: action.payload.userId
                }
            });

            return {
                ...response.body,
            };
        }, action.payload);
    }
}

/**
 * Trigger the callback when loading the user has finished
 * @param action the action to handle
 */
export function loadUserFinishedSaga(action: AsyncAction<LoadUserAction>) {
    if (action.input) {
        action.input.callback();
    }
}

/**
 * Reducer for storing a user
 * @param state the state to update
 * @param action the action to update the state with
 * @return the new state
 */
export function storeUserReducer(state: State, action: ResolvedAsyncAction<LoadUserAction, User>) : State {
    return produce(state, (draft: State) => {
        if (action.input === undefined || action.payload === undefined) {
            return;
        }

        const id = action.payload.id;
        const users = draft.users.filter(user => user.id !== id);

        draft.users = [
            action.payload,
            ...users,
        ];
    });
}

//////// Selectors for getting user data

/** Selector to get the specified user */
export const selectUserById = (id: string) =>
    buildSelector(['users', 'users'], (users: Array<User>) =>
        users.find(user => user.id === id));

//////// The actual module definitions

/** The sagas for this module */
export const sagas = [
    buildSaga(LOAD_USER_ACTION, loadUserSaga),
    buildSaga(finishedAction(LOAD_USER_ACTION), loadUserFinishedSaga),
];

/** The reducers for this module */
export const reducers = createReducer(INITIAL_STATE, {
    [succeededAction(LOAD_USER_ACTION)] : storeUserReducer,
    [succeededAction(UPDATE_PROFILE_ACTION)] : storeUserReducer,
});
