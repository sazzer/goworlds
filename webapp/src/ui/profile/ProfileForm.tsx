import React, {FunctionComponent, useState} from 'react';
import {Button, Form} from "semantic-ui-react";
import {useTranslation} from "react-i18next";
import * as Yup from "yup";
import {Formik} from "formik";
import {FormikErrorMessage} from "../common/FormikErrorMessage";
import {ErrorMessage} from "../common/ErrorMessage";
import {User} from "../../users/users";

/** The props that the ProfileForm area needs */
type ProfileFormProps = {
    user: User,
};

/**
 * Component for editing the core profile details for a user
 * @constructor
 */
export const ProfileForm: FunctionComponent<ProfileFormProps> = ({user}) => {
    const { t } = useTranslation();
    const [submitting, setSubmitting] = useState(false);
    const [error, setError] = useState<string | undefined>(undefined);

    const schema = Yup.object().shape({
        email: Yup.string()
            .required(t('profile.form.email.errors.required'))
            .email(t('profile.form.email.errors.email')),
        name: Yup.string()
            .required(t('profile.form.name.errors.required')),
    });

    const doSubmit = (email: string, name: string) => {
        setSubmitting(true);
        setError(undefined);
    };

    return (
        <Formik initialValues={user}
                validationSchema={schema}
                onSubmit={(values) => doSubmit(values.email, values.name)}>
            {({values, isValid, errors, touched, handleSubmit, handleChange, handleBlur, handleReset, dirty}) =>
                <Form onSubmit={handleSubmit}
                      error={!isValid || error !== undefined}
                      loading={submitting}
                      data-test="ProfileForm">
                    <Form.Field required data-test="email">
                        <label>
                            {t('profile.form.email.label')}
                        </label>
                        <Form.Input fluid
                                    placeholder={t('profile.form.email.placeholder')}
                                    name="email"
                                    type="text"
                                    value={values.email}
                                    error={touched.email && errors.email !== undefined}
                                    onChange={handleChange}
                                    onBlur={handleBlur}
                                    autoFocus/>
                        <FormikErrorMessage name="email" />
                    </Form.Field>
                    <Form.Field required data-test="name">
                        <label>
                            {t('profile.form.name.label')}
                        </label>
                        <Form.Input fluid
                                    placeholder={t('profile.form.name.placeholder')}
                                    name="name"
                                    type="text"
                                    value={values.name}
                                    error={touched.name && errors.name !== undefined}
                                    onChange={handleChange}
                                    onBlur={handleBlur}/>
                        <FormikErrorMessage name="name" />
                    </Form.Field>
                    <Button type="submit" primary disabled={!dirty}>
                        {t('profile.form.submit')}
                    </Button>
                    <Button type="reset" negative onClick={handleReset} disabled={!dirty}>
                        {t('profile.form.reset')}
                    </Button>
                    <ErrorMessage testName="ProfileFormErrors" errors={[
                        error && t('profile.form.submit.errors.' + error)
                    ]}/>
                </Form>
            }
        </Formik>
    );
};
