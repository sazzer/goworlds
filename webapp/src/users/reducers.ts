import {reducers as users} from "./users";

export default {
    users,
};
