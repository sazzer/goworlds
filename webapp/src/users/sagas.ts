import { sagas as usersSagas } from './users';
import { sagas as updateProfileSagas } from './updateProfile';

export default [
    ...usersSagas,
    ...updateProfileSagas,
];
