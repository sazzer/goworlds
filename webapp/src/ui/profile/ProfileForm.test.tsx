import React from 'react';
import {ProfileForm} from './ProfileForm';
import {render, fireEvent} from '@testing-library/react';
import configureStore from "redux-mock-store";
import {Provider} from "react-redux";
import {ProblemError} from "../../api";
import {User} from "../../users/user";

/** Set up the component to test */
function setup(user: User) {
    const mockStoreCreator = configureStore();
    const store = mockStoreCreator({});

    const element = render(<Provider store={store}><ProfileForm user={user} /></Provider>);

    const enterEmail = (email: string) => {
        fireEvent.change(element.getByPlaceholderText('Enter Email Address'), {
            target: {
                value: email,
            }
        });
    };

    const enterName = (email: string) => {
        fireEvent.change(element.getByPlaceholderText('Enter Name'), {
            target: {
                value: email,
            }
        });
    };

    const submitForm = () => new Promise(resolve => {
        fireEvent.click(element.getByText('Save Changes'));
        setImmediate(resolve);
    });

    const resetForm = () => new Promise(resolve => {
        fireEvent.click(element.getByText('Reset Changes'));
        setImmediate(resolve);
    });

    return {
        store,

        element,

        enterEmail,
        enterName,
        submitForm,
        resetForm
    };
}

it('renders the initial state', () => {
    const {element} = setup({
        id: 'abc123',
        name: 'Test User',
        email: 'test@example.com',
        created: '',
        updated: '',
    });

    expect(element.baseElement).toMatchSnapshot();
});

it('renders correctly after entering new details', () => {
    const {element, enterName, enterEmail} = setup({
        id: 'abc123',
        name: 'Test User',
        email: 'test@example.com',
        created: '',
        updated: '',
    });

    enterEmail('new@example.co.uk');
    enterName('New Name');

    expect(element.baseElement).toMatchSnapshot();
});

it('renders correctly after resetting', async () => {
    const {element, enterName, enterEmail, resetForm} = setup({
        id: 'abc123',
        name: 'Test User',
        email: 'test@example.com',
        created: '',
        updated: '',
    });

    enterEmail('new@example.co.uk');
    enterName('New Name');
    await resetForm();

    expect(element.baseElement).toMatchSnapshot();
});

it('renders errors if submitting when blank', async () => {
    const {element, enterName, enterEmail, submitForm} = setup({
        id: 'abc123',
        name: 'Test User',
        email: 'test@example.com',
        created: '',
        updated: '',
    });

    enterEmail('');
    enterName('');
    await submitForm();

    expect(element.baseElement).toMatchSnapshot();
});

it('Marks as loading when submitting valid details', async () => {
    const {element, enterName, enterEmail, submitForm} = setup({
        id: 'abc123',
        name: 'Test User',
        email: 'test@example.com',
        created: '',
        updated: '',
    });

    enterEmail('new@example.co.uk');
    enterName('New Name');
    await submitForm();

    expect(element.baseElement).toMatchSnapshot();
});

it('Triggers the correct redux action when submitting valid details', async () => {
    const {store, enterName, enterEmail, submitForm} = setup({
        id: 'abc123',
        name: 'Test User',
        email: 'test@example.com',
        created: '',
        updated: '',
    });

    enterEmail('new@example.co.uk');
    enterName('New Name');
    await submitForm();

    expect(store.getActions()).toHaveLength(1);
    expect(store.getActions()[0]).toMatchObject({
        type: 'Users/updateProfile',
        payload: {
            userId: 'abc123',
            name: 'New Name',
            email: 'new@example.co.uk'
        }
    });
});

it('Renders correctly when submitting triggers the callback as success', async () => {
    const {element, store, enterName, enterEmail, submitForm} = setup({
        id: 'abc123',
        name: 'Test User',
        email: 'test@example.com',
        created: '',
        updated: '',
    });

    enterEmail('new@example.co.uk');
    enterName('New Name');
    await submitForm();

    expect(store.getActions()).toHaveLength(1);
    store.getActions()[0].payload.callback(undefined);

    expect(element.baseElement).toMatchSnapshot();
});

it('Renders correctly when submitting triggers the callback as an unexpected failure', async () => {
    const {element, store, enterName, enterEmail, submitForm} = setup({
        id: 'abc123',
        name: 'Test User',
        email: 'test@example.com',
        created: '',
        updated: '',
    });

    enterEmail('new@example.co.uk');
    enterName('New Name');
    await submitForm();

    expect(store.getActions()).toHaveLength(1);
    store.getActions()[0].payload.callback(new Error());

    expect(element.baseElement).toMatchSnapshot();
});

it('Renders correctly when submitting triggers the callback as a duplicate email failure', async () => {
    const {element, store, enterName, enterEmail, submitForm} = setup({
        id: 'abc123',
        name: 'Test User',
        email: 'test@example.com',
        created: '',
        updated: '',
    });

    enterEmail('new@example.co.uk');
    enterName('New Name');
    await submitForm();

    expect(store.getActions()).toHaveLength(1);
    store.getActions()[0].payload.callback(new ProblemError('',
        400,
        'tag:goworlds,2019:users/problems/duplicate-email-address',
        null));

    expect(element.baseElement).toMatchSnapshot();
});