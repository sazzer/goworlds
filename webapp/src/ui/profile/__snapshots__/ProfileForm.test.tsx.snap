// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Marks as loading when submitting valid details 1`] = `
<body>
  <div>
    <form
      class="ui loading form"
      data-test="ProfileForm"
    >
      <div
        class="required field"
        data-test="email"
      >
        <label>
          Email Address
        </label>
        <div
          class="field"
        >
          <div
            class="ui fluid input"
          >
            <input
              name="email"
              placeholder="Enter Email Address"
              type="text"
              value="new@example.co.uk"
            />
          </div>
        </div>
      </div>
      <div
        class="required field"
        data-test="name"
      >
        <label>
          Name
        </label>
        <div
          class="field"
        >
          <div
            class="ui fluid input"
          >
            <input
              name="name"
              placeholder="Enter Name"
              type="text"
              value="New Name"
            />
          </div>
        </div>
      </div>
      <button
        class="ui primary button"
        type="submit"
      >
        Save Changes
      </button>
      <button
        class="ui negative button"
        type="reset"
      >
        Reset Changes
      </button>
      <div
        class="ui error message"
        data-test="ProfileFormErrors"
      />
    </form>
  </div>
</body>
`;

exports[`Renders correctly when submitting triggers the callback as a duplicate email failure 1`] = `
<body>
  <div>
    <form
      class="ui error form"
      data-test="ProfileForm"
    >
      <div
        class="required field"
        data-test="email"
      >
        <label>
          Email Address
        </label>
        <div
          class="field"
        >
          <div
            class="ui fluid input"
          >
            <input
              name="email"
              placeholder="Enter Email Address"
              type="text"
              value="new@example.co.uk"
            />
          </div>
        </div>
      </div>
      <div
        class="required field"
        data-test="name"
      >
        <label>
          Name
        </label>
        <div
          class="field"
        >
          <div
            class="ui fluid input"
          >
            <input
              name="name"
              placeholder="Enter Name"
              type="text"
              value="New Name"
            />
          </div>
        </div>
      </div>
      <button
        class="ui primary button"
        type="submit"
      >
        Save Changes
      </button>
      <button
        class="ui negative button"
        type="reset"
      >
        Reset Changes
      </button>
      <div
        class="ui error message"
        data-test="ProfileFormErrors"
      >
        Email Address is already registered
      </div>
    </form>
  </div>
</body>
`;

exports[`Renders correctly when submitting triggers the callback as an unexpected failure 1`] = `
<body>
  <div>
    <form
      class="ui error form"
      data-test="ProfileForm"
    >
      <div
        class="required field"
        data-test="email"
      >
        <label>
          Email Address
        </label>
        <div
          class="field"
        >
          <div
            class="ui fluid input"
          >
            <input
              name="email"
              placeholder="Enter Email Address"
              type="text"
              value="new@example.co.uk"
            />
          </div>
        </div>
      </div>
      <div
        class="required field"
        data-test="name"
      >
        <label>
          Name
        </label>
        <div
          class="field"
        >
          <div
            class="ui fluid input"
          >
            <input
              name="name"
              placeholder="Enter Name"
              type="text"
              value="New Name"
            />
          </div>
        </div>
      </div>
      <button
        class="ui primary button"
        type="submit"
      >
        Save Changes
      </button>
      <button
        class="ui negative button"
        type="reset"
      >
        Reset Changes
      </button>
      <div
        class="ui error message"
        data-test="ProfileFormErrors"
      >
        An unexpected error occurred
      </div>
    </form>
  </div>
</body>
`;

exports[`Renders correctly when submitting triggers the callback as success 1`] = `
<body>
  <div>
    <form
      class="ui form"
      data-test="ProfileForm"
    >
      <div
        class="required field"
        data-test="email"
      >
        <label>
          Email Address
        </label>
        <div
          class="field"
        >
          <div
            class="ui fluid input"
          >
            <input
              name="email"
              placeholder="Enter Email Address"
              type="text"
              value="new@example.co.uk"
            />
          </div>
        </div>
      </div>
      <div
        class="required field"
        data-test="name"
      >
        <label>
          Name
        </label>
        <div
          class="field"
        >
          <div
            class="ui fluid input"
          >
            <input
              name="name"
              placeholder="Enter Name"
              type="text"
              value="New Name"
            />
          </div>
        </div>
      </div>
      <button
        class="ui primary button"
        type="submit"
      >
        Save Changes
      </button>
      <button
        class="ui negative button"
        type="reset"
      >
        Reset Changes
      </button>
      <div
        class="ui error message"
        data-test="ProfileFormErrors"
      />
      <div
        class="ui positive message"
      >
        User Profile Updated
      </div>
    </form>
  </div>
</body>
`;

exports[`renders correctly after entering new details 1`] = `
<body>
  <div>
    <form
      class="ui form"
      data-test="ProfileForm"
    >
      <div
        class="required field"
        data-test="email"
      >
        <label>
          Email Address
        </label>
        <div
          class="field"
        >
          <div
            class="ui fluid input"
          >
            <input
              name="email"
              placeholder="Enter Email Address"
              type="text"
              value="new@example.co.uk"
            />
          </div>
        </div>
      </div>
      <div
        class="required field"
        data-test="name"
      >
        <label>
          Name
        </label>
        <div
          class="field"
        >
          <div
            class="ui fluid input"
          >
            <input
              name="name"
              placeholder="Enter Name"
              type="text"
              value="New Name"
            />
          </div>
        </div>
      </div>
      <button
        class="ui primary button"
        type="submit"
      >
        Save Changes
      </button>
      <button
        class="ui negative button"
        type="reset"
      >
        Reset Changes
      </button>
      <div
        class="ui error message"
        data-test="ProfileFormErrors"
      />
    </form>
  </div>
</body>
`;

exports[`renders correctly after resetting 1`] = `
<body>
  <div>
    <form
      class="ui error form"
      data-test="ProfileForm"
    >
      <div
        class="required field"
        data-test="email"
      >
        <label>
          Email Address
        </label>
        <div
          class="field"
        >
          <div
            class="ui fluid input"
          >
            <input
              name="email"
              placeholder="Enter Email Address"
              type="text"
              value="test@example.com"
            />
          </div>
        </div>
      </div>
      <div
        class="required field"
        data-test="name"
      >
        <label>
          Name
        </label>
        <div
          class="field"
        >
          <div
            class="ui fluid input"
          >
            <input
              name="name"
              placeholder="Enter Name"
              type="text"
              value="Test User"
            />
          </div>
        </div>
      </div>
      <button
        class="ui primary disabled button"
        disabled=""
        tabindex="-1"
        type="submit"
      >
        Save Changes
      </button>
      <button
        class="ui negative disabled button"
        disabled=""
        tabindex="-1"
        type="reset"
      >
        Reset Changes
      </button>
      <div
        class="ui error message"
        data-test="ProfileFormErrors"
      />
    </form>
  </div>
</body>
`;

exports[`renders errors if submitting when blank 1`] = `
<body>
  <div>
    <form
      class="ui error form"
      data-test="ProfileForm"
    >
      <div
        class="required field"
        data-test="email"
      >
        <label>
          Email Address
        </label>
        <div
          class="error field"
        >
          <div
            class="ui fluid input"
          >
            <input
              name="email"
              placeholder="Enter Email Address"
              type="text"
              value=""
            />
          </div>
        </div>
        <div
          class="ui error message"
        >
          Please enter an Email Address
        </div>
      </div>
      <div
        class="required field"
        data-test="name"
      >
        <label>
          Name
        </label>
        <div
          class="error field"
        >
          <div
            class="ui fluid input"
          >
            <input
              name="name"
              placeholder="Enter Name"
              type="text"
              value=""
            />
          </div>
        </div>
        <div
          class="ui error message"
        >
          Please enter a user name
        </div>
      </div>
      <button
        class="ui primary button"
        type="submit"
      >
        Save Changes
      </button>
      <button
        class="ui negative button"
        type="reset"
      >
        Reset Changes
      </button>
      <div
        class="ui error message"
        data-test="ProfileFormErrors"
      />
    </form>
  </div>
</body>
`;

exports[`renders the initial state 1`] = `
<body>
  <div>
    <form
      class="ui error form"
      data-test="ProfileForm"
    >
      <div
        class="required field"
        data-test="email"
      >
        <label>
          Email Address
        </label>
        <div
          class="field"
        >
          <div
            class="ui fluid input"
          >
            <input
              name="email"
              placeholder="Enter Email Address"
              type="text"
              value="test@example.com"
            />
          </div>
        </div>
      </div>
      <div
        class="required field"
        data-test="name"
      >
        <label>
          Name
        </label>
        <div
          class="field"
        >
          <div
            class="ui fluid input"
          >
            <input
              name="name"
              placeholder="Enter Name"
              type="text"
              value="Test User"
            />
          </div>
        </div>
      </div>
      <button
        class="ui primary disabled button"
        disabled=""
        tabindex="-1"
        type="submit"
      >
        Save Changes
      </button>
      <button
        class="ui negative disabled button"
        disabled=""
        tabindex="-1"
        type="reset"
      >
        Reset Changes
      </button>
      <div
        class="ui error message"
        data-test="ProfileFormErrors"
      />
    </form>
  </div>
</body>
`;
